src/passes/RemoveUnusedModuleElements.cpp

```diff
@@ -313,28 +323,34 @@ struct RemoveUnusedModuleElements : public Pass {
       roots.emplace_back(ModuleElementKind::Function, name);
     });
     // Compute reachability starting from the root set.
-    ReachabilityAnalyzer analyzer(module, roots);
+    auto closedWorld = getPassOptions().closedWorld;
+    ReachabilityAnalyzer analyzer(module, roots, closedWorld);
 
     // RefFuncs that are never called are a special case: We cannot remove the
     // function, since then (ref.func $foo) would not validate. But if we know
     // it is never called, at least the contents do not matter, so we can
     // empty it out.
+    //
+    // We can only do this in a closed world, as otherwise function references
+    // may be called outside of the module (if they escape, which we could in
+    // principle track, see the TODO earlier in this file).
     std::unordered_set<Name> uncalledRefFuncs;
-    for (auto& [type, targets] : analyzer.uncalledRefFuncMap) {
-      for (auto target : targets) {
-        uncalledRefFuncs.insert(target);
-      }
+    if (closedWorld) {
```

Comment:
Aren't we adding anything anyway if `closedWorld` is not true, in `ReachabilityAnalyzer ::visitCallRef`?

Reply:
Good point, yeah, `uncalledRefFuncMap` will be empty then. I removed the `if` (and added an assert).